test/src/e2e_vm_tests/test_programs/should_fail/match_expressions_or_mismatched_types/src/main.sw

```diff
@@ -0,0 +1,239 @@
+script;
+
+mod lib;
+
+use ::lib::LibStruct;
+use ::lib::LibStructAlias;
+
+use ::lib::LibStruct as LibStructPlayAlias;
+use ::lib::LibStructAlias as LibStructAliasPlayAlias;
+
+struct Struct {
+    x: bool,
+    y: u64,
+}
+
+impl Struct {
+    fn new() -> Self {
+        Self {
+            x: false,
+            y: 0,
+        }
+    }
+
+    fn use_me(self) -> () {
+        poke(self.x);
+        poke(self.y);
+    }
+}
+
+struct GenStruct<A, B> {
+    x: A,
+    y: B,
+}
+
+impl<A, B> GenStruct<A, B> {
+    fn new(a: A, b: B) -> Self {
+        Self {
+            x: a,
+            y: b,
+        }
+    }
+
+    fn use_me(self) -> () {
+        poke(self.x);
+        poke(self.y);
+    }
+}
+
+type StructAlias = Struct;
+type GenStructAliasBool = GenStruct<bool, bool>;
+type Boolean = bool;
+type GenStructAliasBoolean = GenStruct<Boolean, Boolean>;
+
+trait Supertrait { }
+trait Subtrait: Supertrait { }
+
+impl Supertrait for Struct { }
+impl Subtrait for Struct { }
+
+enum Enum {
+    A: (bool),
+    B: (u64),
+    C: (Struct),
+    D: (StructAlias),
+    E: (GenStruct<bool, bool>),
+    F: (GenStructAliasBool),
+    G: (GenStructAliasBoolean),
+    H: (LibStruct),
+    I: (LibStructAlias),
+    J: (LibStructPlayAlias),
+    K: (LibStructAliasPlayAlias),
+}
+
+fn generics<A, B>(a: A, b: B)
+    where A: Supertrait,
+          B: Subtrait
+{
+    let gen = (a, b);
+
+    let _x = match gen {
+        (x, y) | (y, x) => 0,
+    };
+}
+
+fn main() {
+    let s1 = Struct::new();
+
+    let _x = match s1 {
+        Struct { x, y } | Struct { x: y, y: x } => if y { 0 } else { x },
```

Comment:
Should this be accepted?
At `Struct { x, y }`, `y` is `u64`; but at `Struct { x: y, y: x }`, `y` is `bool`. Rustc correctly complains.

```
   |
17 |     let _x = match s1 {
   |                    -- this expression has type `Struct`
18 |         Struct { x, y } | Struct { x: y, y: x } => if y { 0 } else { x },
   |                     -                 ^ expected `u64`, found `bool`
   |                     |
   |                     first introduced with type `u64` here
   |
   = note: in the same arm, a binding must have the same type in all alternatives
```

Reply:
Good point! The existing code was short-circuiting at the first inconsistency between the checked and the previous alternative and I kept that logic intact. But it's definitely worth questioning, both for this new error and the existing one (which was buggy and also fixed). The question is only if we don't short-circuit immediately, how far do we go. Do we collect all inconsistencies in all the alternatives or only all in the currently checked and then short-circuit?

The latter approach should be straightforward and should also emit error in the case you are pointing.

I have to see if the former approach potentially clashes with the current logic in which (unlike Rust!) the last alternative wins. I personally find that approach confusing because intuitively I see the first alternative as the "source of truth" that others should follow.

Anyhow, putting the PR back to draft until I give it a thought and decide ;-) Thanks!

Reply:
Gave it a second thought :-) I definitely want to emit as much as possible but let's do it in a separate PR that will address #5108 and leave this implementation as is.

Essentially, I want to change the original algorithm (remove redundant `if` and `Some`, make the first alternative as the reference and not the ever-changing current one, ...), double check the affected code areas and existing tests, and also use the opportunity to move the existing error for non-bound variable to expressive diagnostics.

I've commented in the PR description that the current algorithm short-circuits.